x-pack/filebeat/input/salesforce/input.go

```diff
@@ -0,0 +1,506 @@
+			if err != nil {
+				return err
+			}
+
+			body, err := io.ReadAll(resp.Body)
+			if err != nil {
+				resp.Body.Close()
+				return err
+			}
+			resp.Body.Close()
+
+			recs, err := decodeAsCSV(body)
+			if err != nil {
+				return err
+			}
+
+			if firstEvent {
+				if timstamp, ok := rec.Record().Fields()[s.config.EventMonitoringMethod.EventLogFile.Cursor.Field].(string); ok {
+					s.cursor.EventLogFile.FirstEventTime = timstamp
+				}
+			}
+
+			if timstamp, ok := rec.Record().Fields()[s.config.EventMonitoringMethod.EventLogFile.Cursor.Field].(string); ok {
+				s.cursor.EventLogFile.LastEventTime = timstamp
+			}
+
+			for _, val := range recs {
+				jsonStrEvent, err := json.Marshal(val)
+				if err != nil {
+					return err
+				}
+
+				err = publishEvent(s.publisher, s.cursor, jsonStrEvent, "EventLogFile")
+				if err != nil {
+					return err
+				}
+				totalEvents++
+			}
+			firstEvent = false
+		}
+
+		if res.MoreRecords() {
+			res, err = res.Next()
+			if err != nil {
+				return err
+			}
+		} else {
+			break
+		}
+	}
+	s.log.Debugf("Total events: %d", totalEvents)
+
+	return nil
+}
+
+// getSFDCConfig returns a new Salesforce configuration based on the configuration.
+func getSFDCConfig(cfg *config) (*sfdc.Configuration, error) {
+	var (
+		creds *credentials.Credentials
+		err   error
+	)
+
+	if cfg.Auth == nil {
+		return nil, errors.New("no auth provider enabled")
+	}
+
+	switch {
+	case cfg.Auth.OAuth2.JWTBearerFlow != nil && cfg.Auth.OAuth2.JWTBearerFlow.isEnabled():
+		pemBytes, err := os.ReadFile(cfg.Auth.OAuth2.JWTBearerFlow.ClientKeyPath)
+		if err != nil {
+			return nil, fmt.Errorf("problem with client key path for JWT auth: %w", err)
+		}
+
+		signKey, err := jwt.ParseRSAPrivateKeyFromPEM(pemBytes)
+		if err != nil {
+			return nil, fmt.Errorf("problem with client key for JWT auth: %w", err)
+		}
+
+		passCreds := credentials.JwtCredentials{
+			URL:            cfg.Auth.OAuth2.JWTBearerFlow.URL,
+			ClientId:       cfg.Auth.OAuth2.JWTBearerFlow.ClientID,
+			ClientUsername: cfg.Auth.OAuth2.JWTBearerFlow.ClientUsername,
+			ClientKey:      signKey,
+		}
+
+		creds, err = credentials.NewJWTCredentials(passCreds)
+		if err != nil {
+			return nil, fmt.Errorf("problem with credentials: %w", err)
+		}
+	case cfg.Auth.OAuth2.UserPasswordFlow != nil && cfg.Auth.OAuth2.UserPasswordFlow.isEnabled():
+		passCreds := credentials.PasswordCredentials{
+			URL:          cfg.Auth.OAuth2.UserPasswordFlow.TokenURL,
+			Username:     cfg.Auth.OAuth2.UserPasswordFlow.Username,
+			Password:     cfg.Auth.OAuth2.UserPasswordFlow.Password,
+			ClientID:     cfg.Auth.OAuth2.UserPasswordFlow.ClientID,
+			ClientSecret: cfg.Auth.OAuth2.UserPasswordFlow.ClientSecret,
+		}
+
+		creds, err = credentials.NewPasswordCredentials(passCreds)
+		if err != nil {
+			return nil, fmt.Errorf("problem with credentials: %w", err)
+		}
+
+	}
+
+	return &sfdc.Configuration{
+		Credentials: creds,
+		Client:      http.DefaultClient,
```

Comment:
does this mean that timeouts are not configurable for the salesforce API calls?

Reply:
This is a good point. The answer to your question is yes.

Do you have any suggestions for some sane defaults for Timeouts, and KeepAlive?

Meanwhile, let me take a look at some other Salesforce clients; and what defaults they are using.

Reply:
Good enough defaults: https://github.com/golang/go/blob/a56834922fb0acfede96aa2ff19d703d022f391b/src/net/http/transport.go#L43?

Reply:
Also, should we make it configurable by user?

Reply:
i think having this be user-configurable would be good. for defaults, i'm not sure, does SFDC have any sample code ?

i believe the default here is never timeout (?), so having some sane default would be essential.